### PR TITLE
Get voxel coordinates with volume affine

### LINKED ISSUE #252: Extrinsic passed to the DRR module
I am not sure what is the extrinsic passed to the DRR module.

It seems that is `isocenter_from_camera`. Usually, extrisnics are from some world frame to the camera frame. But I believe the DRR module takes the inverse?

Also in `v0.4`, I believe the world is at the isocenter of the CT space. It get's pretty hard to do anything with extrinsics in RAS space, since the conversion is lost is in `canocalize` when `image.affine` is overwritten.

### LINKED ISSUE #261: bug in perspective_projection
Thank you for the contribution . I noticed a strange phenomenon in the perspective_projection function of DRR. To verify this issue, I created six spheres under CT and then used the DRR algorithm for projection. When comparing the output of the DRR projection and the perspective_projection, I found significant differences when the angle changes. As shown in the figure, the bright spots represent the projection results, and the circles indicate the estimated positions from the perspective_projection.

When I use this code:
```
         rotations = torch.tensor([[ 1.0,  0., 1]], device=self.device)
         translations = torch.tensor([[ 0, 800., 0.]], device=self.device)
                        
        pred_img = drr_seg(rotations, translations, parameterization="euler_angles", convention="ZXY")
        poseIV = convert(rotations, translations, parameterization="euler_angles", convention="ZXY")
        points2d =  drr_seg.perspective_projection(poseIV,point3d)
```

id_0_gt_bone_drr

When I use this code:
```
         rotations = torch.tensor([[ .0,  0., 0]], device=self.device)
         translations = torch.tensor([[ 0, 800., 0.]], device=self.device)
                        
        pred_img = drr_seg(rotations, translations, parameterization="euler_angles", convention="ZXY")
        poseIV = convert(rotations, translations, parameterization="euler_angles", convention="ZXY")
        points2d =  drr_seg.perspective_projection(poseIV,point3d)
```
id_0_gt_bone_drr

### PR DESCRIPTION
**Pull request by coolteemf**
Fix #261 by removing call to `ToCanonical` class in canonicalize function and using volume affine matrix in the renderer instead. Also closes #252.

Naturally handles non-standard volume affines, in order to be more compatible with other frameworks like DeepDRR and 3DSlicer.

### TODO

- [x] Implement forward model with affine matrix
- [x] Figure out `torch.float32` vs `torch.float64`
- [x] Rerun and reproduce `introduction.ipynb`
- [x]  Rerun and reproduce `geometry.ipynb`
- [x] Rerun and reproduce `registration.ipynb` (albeit, with `trilinear`, not `siddon`)
- [x] Rerun and reproduce `reconstruction.ipynb`
- [x] Rerun and reproduce `metrics.ipynb`
- [x] Rerun and reproduce `timing.ipynb`
- [x] Rerun and reproduce `trilinear.ipynb`

### CODE REVIEW

**Comment by eigenvivek** [comment]
Hi @coolteemf , you need to run `nbdev_clean; nbdev_export` to compile the python source files.

I've tried out the code in a few places and it looks like it's working really well! I'll keep running the rest of the tutorial notebooks.

**Comment by coolteemf** [comment]
Ok done.
Thanks, I hope there are no bugs !

File: diffdrr/renderers.py [hunk 1]
```diff
@@ -90,21 +88,30 @@ def forward(
...
     # Parameterize the parallel XYZ planes that comprise the CT volumes
-    alphax = (torch.arange(dims[0] + 1).to(source) - 0.5) * spacing[0] + origin[0]
-    alphay = (torch.arange(dims[1] + 1).to(source) - 0.5) * spacing[1] + origin[1]
-    alphaz = (torch.arange(dims[2] + 1).to(source) - 0.5) * spacing[2] + origin[2]
+    alphax = torch.arange(dims[0] + 1).to(source) - 0.5
+    alphay = torch.arange(dims[1] + 1).to(source) - 0.5
+    alphaz = torch.arange(dims[2] + 1).to(source) - 0.5
 
     # Calculate the parametric intersection of each ray with every plane
+    source = (
```
Comment (eigenvivek) [inline diffdrr/renderers.py:107]: Can probably rewrite this using the RigidPose class.

**Comment by eigenvivek** [comment]
Not sure why the tests are failing, i’ll figure it out later today.

**Comment by haarisr** [comment]
> Not sure why the tests are failing, i’ll figure it out later today.

Error's gotta do with numpy 2.0. If you pin the numpy version to 1.x.x it should pass.

**Comment by eigenvivek** [comment]
@haarisr Thanks that was indeed the fix! We can remove the strict versioning once `torchio` updates to support `torch=2.3`.

File: diffdrr/renderers.py [hunk 2]
```diff
@@ -90,21 +88,30 @@ def forward(
...
+        torch.einsum("nab, n...b -> n...a", inverse_affine[:, :3, :3], source)
+        + inverse_affine[:, :3, 3][:, None]
+    )
+    target = (
+        torch.einsum("nab, n...b -> n...a", inverse_affine[:, :3, :3], target)
+        + inverse_affine[:, :3, 3][:, None]
+    )
```
Comment (eigenvivek) [inline diffdrr/renderers.py:114]: More clean to implement `affine` and `inverse_affine` as `diffdrr.pose.RigidTransform`s.

**Comment by haarisr** [comment]
> I've tried out the code in a few places and it looks like it's working really well! I'll keep running the rest of the tutorial notebooks.

Does this work on the registration notebook? The runs do not seem to converge?

**Comment by coolteemf** [comment]
> Does this work on the registration notebook? The runs do not seem to converge?

I'm experiencing issues with registration on my data too, and I'm not sure where the issue is.
I will investigate tomorrow.

**Comment by eigenvivek** [comment]
I've noticed that the registration tutorial is very brittle to changes in the rendering algorithm (i.e., hyperparameters need to be tuned again). It's a strange issue since the gradients should be (in theory) the same, but tweaking the learning rate _should_ fix the issue.

**Comment by coolteemf** [comment]
I haven't pulled the latest changes, but I think I found a bug here:
```
@patch
def forward(self: Detector, extrinsic: RigidTransform, calibration: RigidTransform):
    """Create source and target points for X-rays to trace through the volume."""
    if calibration is None:
        target = self.calibration(self.target)
    else:
        target = calibration(self.target)
    pose = self.reorient.compose(extrinsic)
    source = pose(self.source)
    target = pose(target)
    return source, target
```
When I use a pose without rotation component (ie simple matrix with just a translation in the pose), it works and source - target norm is between 1100 and 1120
When I use another pose, the source - target norm is between 2251 and 2510, which is not normal since I didn't change the SDD.
And I'm not sure whats wrong, since the difference between source and target is just that calibration is applied to target, which should not depend on pose.

**Comment by coolteemf** [comment]
Just deleted my comment, it was a mistake in my code.
Still, I think something may be wrong since I can't convert between deepdrr and diffdrr extrinsic matrices anymore.
I'm continuing to look into it.

**Comment by coolteemf** [comment]
I think I found the issue and I'm not sure it has something to do with diffdrr.
get_alphas sample alphas between 0 and 1, and my pose, for some reason, is looking at the volume but from too far and so no points get sampled in the volume.
I plotted the "xyz" coordinates after get_xyz and befor normalization.
I would say the code in this PR is safe (from this bug at least), but I won't be too sure until I've solved the bug.
image

**Comment by coolteemf** [comment]
> Just deleted my comment, it was a mistake in my code. Still, I think something may be wrong since I can't convert between deepdrr and diffdrr extrinsic matrices anymore. I'm continuing to look into it.

Converting between frameworks still works, and I got registration to work too.
I think the code is fine.

**Comment by eigenvivek** [comment]
Great! Can you push your version, along with the updated registration notebook? I'll try to reproduce it on my machine. I've found that sometimes the registration tutorial will run on one of my GPUs, but not on another (different models).

Also how are you converting between DeepDRR and DiffDRR? Is it the same as the `metamorphasis` notebook (would be surprised if so because I haven't updated this in a year lol)? Perhaps this would be better for another PR, but if you have a good method for converting between the two renderers, it would be amazing to have an updated tutorial on how to do that.

**Comment by haarisr** [comment]
Side note

Torchio updated their dependencies so we can revert the pinned versions in this PR.
https://github.com/fepegar/torchio/pull/1183

**Comment by coolteemf** [comment]
> Great! Can you push your version, along with the updated registration notebook? I'll try to reproduce it on my machine. I've found that sometimes the registration tutorial will run on one of my GPUs, but not on another (different models).

Sorry I meant that I made registration work on my own data. 
I don't know why the registration tutorial has different results between GPUs.
I am suspecting that it is related to precision differences with the use of tensor cores on some GPUs.
I wonder if the use of double precision in the affine also causes discrepancies.
If we could use float instead with some trick it would be great, because double are unsupported on apple silicon GPUs...

> Also how are you converting between DeepDRR and DiffDRR?

I'm directly converting from DiffDRR pose matrix to DeepDRR extrinsic matrix, taking into account the affine changes that is introduced in the read function of DiffDRR.
In DeepDRR like in metamorphasis, it is rather a MobileCarm that is used and I would have to convert the frame of reference change matrix to motion of the MobileCarm to stay consistent with that.
I'll have a look at how to do that.

**Comment by haarisr** [comment]
Found a way to simplify this PR. 

More details here https://github.com/coolteemf/DiffDRR/pull/1

**Comment by coolteemf** [comment]
While testing @haarisr code, I found a bug in batched rendering. Will try to fix it, and also run nbdev nbexport as well.

**Comment by coolteemf** [comment]
Found the bug.
When applying a batch size 1 RigidTransform (affine_inverse) to a batch size N point cloud (source, target in batched rendering), the result is incorrect.
The part to fix is in this code, unless we repeat affine_inverse to match batch_size.
I don't get what it means for b and k and i and n to collapse in this einsum, can you explain @eigenvivek ?
I don't want to change the expected behavior of RigidTransform in the batched case.
´´´
    def forward(self, x):
        """Apply (a batch) of rigid transforms to a pointcloud."""
        x_pad = torch.nn.functional.pad(x, (0, 1), value=1.0)
        return torch.einsum("bij, knj -> bni", self.matrix[:, :3], x_pad)
´´´

**Comment by eigenvivek** [comment]
> I don't get what it means for b and k and i and n to collapse in this einsum

the `k` is misleading here as I'm pretty sure it's always 1 in every pointcloud I use.

`b` is the number of rigid transforms
`n` is the number of points in the point cloud

**Comment by eigenvivek** [comment]
Just ran the benchmarks in the `timing.ipynb` notebook

For a 200 x 200 DRR, the rendering speed has increased from `27.3 ms -> 48.9 ms`

Not sure why the rendering speed has increased 2X, but my best guess would be that it's coming from the dtype conversions?

**Comment by haarisr** [comment]
> Not sure why the rendering speed has increased 2X, but my best guess would be that it's coming from the dtype conversions?

Yeah the affine being set to float64 in drr module slows it down. I see it to float32 and it was faster. 

I didn't see a huge difference in renewing quality but we should do some tests to see whether it actually improves quality

**Comment by eigenvivek** [comment]
You're right, setting the entire DRR module to `torch.float32` gets a runtime of `24.5 ms` (perhaps even a modest improvement!). I'll see if rendering quality is perceivably different.

Also, can confirm that the `registration.ipynb` notebook does not work at the moment (fails to converge to global max). This notebook broke in a similar way when I changed Siddon's method to be implemented with `grid_sample` (#272), so I think I'll be able to figure out how to fix it.

**Comment by eigenvivek** [comment]
There seems to be very minimal differences between 32-bit and 64-bit DRRs (i.e., the aliasing issues in #202 don't seem to be a problem anymore). So for now, I rolled back all explicit type-setting. By default, everything in the buffer is initialized as `torch.float32`.

download

```python
import matplotlib.pyplot as plt
import torch

from diffdrr.data import load_example_ct
from diffdrr.drr import DRR
from diffdrr.pose import convert
from diffdrr.visualization import plot_drr

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

subject = load_example_ct()

rotations = torch.tensor([[0.0, 0.0, 0.0]], device=device)
translations = torch.tensor([[0.0, 850.0, 0.0]], device=device)
pose = convert(
    rotations, translations, parameterization="euler_angles", convention="ZXY"
)

# 32-bit DRR
drr = DRR(subject, sdd=1020.0, height=200, delx=2.0).to(
    device=device, dtype=torch.float32
)
img1 = drr(pose)

# 64-bit DRR
drr = DRR(subject, sdd=1020.0, height=200, delx=2.0).to(
    device=device, dtype=torch.float64
)
img2 = drr(pose.to(torch.float64))

plot_drr(
    torch.concat([img1, img2, img2 - img1]),
    title=["torch.float32", "torch.float64", "difference"],
)
plt.show()
```

**Comment by haarisr** [comment]
Id suggest keeping things by default to float32 and let the user pass in a dtype if he wanted to

**Comment by eigenvivek** [comment]
In the 3D rendering functions, we have to use `ToCanonical` since it seems pyvista expects voxelgrids to be in RAS+ orientation. I guess this means that we don't yet support 3D plotting of non-standard volumes, but I think that is an issue for future PR.

**Comment by haarisr** [comment]
> In the 3D rendering functions, we have to use `ToCanonical` since it seems pyvista expects voxelgrids to be in RAS+ orientation. I guess this means that we don't yet support 3D plotting of non-standard volumes, but I think that is an issue for future PR.

Was just playing around and found that this is also similar. I am not sure if it would work all the time

```
    from torchio.data.io import get_rotation_and_spacing_from_affine

    rotation, spacing = get_rotation_and_spacing_from_affine(subject.volume.affine)
    origin = rotation @ subject.volume.origin
    spacing = subject.volume.spacing

```
You then pass this spacing and  origin to the visualzation functions

**Comment by eigenvivek** [comment]
As far as I can tell, the gradients flowing through Siddon's method are broken for registration.

In `registration.ipynb`, to get the image similarity metric to increase, I need to change `maximize=False` in the optimizer. That is, the gradients appear to be incorrectly pointing in the direction of steepest *descent*, for some mystical reason. Even after setting `maximize=False`, the image similarity metric maxes out at around 0.95 out of 1.

This hack isn't needed when using `renderer="trilinear"`. The notebook works without issue with this backend (image similarity metric reaches 0.999 very quickly). 

Siddon's method still works perfectly fine for `reconstruction.ipynb`, so I'm not sure what the issue is.

**Comment by eigenvivek** [comment]
> Was just playing around and found that this is also similar. I am not sure if it would work all the time
> 
> ```
>     from torchio.data.io import get_rotation_and_spacing_from_affine
> 
>     rotation, spacing = get_rotation_and_spacing_from_affine(subject.volume.affine)
>     origin = rotation @ subject.volume.origin
>     spacing = subject.volume.spacing
> ```
> 
> You then pass this spacing and origin to the visualzation functions

@haarisr Just thought of one other idea - PyVista lets you transform meshes with 4x4 matrices (see [here](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.transform.html)). Perhaps we could initialize the `ImageData` with `origin=(0, 0, 0)` and `spacing=(1, 1, 1)`, then transform it with the volume's affine matrix.

Probably good to implement in a new issue.

**Comment by eigenvivek** [comment]
> As far as I can tell, the gradients flowing through Siddon's method are broken for registration.

In the interest of getting this PR merged, I propose to change the default renderer to `trilinear` and add a warning to Siddon's method saying that it's gradients are currently unstable for registration. However, before releasing a new version of DiffDRR, I'd like to resolve whatever issue currently plagues Siddon's gradients.

Let me know if you think something else might be better.

**Comment by haarisr** [comment]
> In `registration.ipynb`, to get the image similarity metric to increase, I need to change `maximize=False` in the optimizer. That is, the gradients appear to be incorrectly pointing in the direction of steepest _descent_, for some mystical reason. Even after setting `maximize=False`, the image similarity metric maxes out at around 0.95 out of 1.

That's wild. Curious how you debugged this.

**Comment by coolteemf** [comment]
> As far as I can tell, the gradients flowing through Siddon's method are broken for registration.
> 
> In `registration.ipynb`, to get the image similarity metric to increase, I need to change `maximize=False` in the optimizer. That is, the gradients appear to be incorrectly pointing in the direction of steepest _descent_, for some mystical reason. Even after setting `maximize=False`, the image similarity metric maxes out at around 0.95 out of 1.
> 
> This hack isn't needed when using `renderer="trilinear"`. The notebook works without issue with this backend (image similarity metric reaches 0.999 very quickly).
> 
> Siddon's method still works perfectly fine for `reconstruction.ipynb`, so I'm not sure what the issue is. Current best guess is that `torch.nn.functional.grid_sample` w/ `mode="nearest"` has broken something that previously worked with the custom implementation of nearest-neighbor interpolation?
> 
> 🤷🏾‍♀️

Yes sampling mode nearest is not really differentiable.
Gradients are not continuous with this interpolation, so I wouldn't use it in a differentiable function.
Had the same problem when optimizing for best DICE of deformed segmentation, I replaced sampling mode with bilinear and it worked perfectly and I binarized after interpolation to obtain valid segmentation.
